File: common/src/main/kotlin/org/sandboxpowered/silica/registry/RegistryObject.java
package org.sandboxpowered.silica.registry;

import org.sandboxpowered.silica.util.Identifier;

import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.function.Supplier;

public interface RegistryObject<T extends RegistryEntry<T>> extends Supplier<T> {
    Optional<T> asOptional();

    T get() throws NoSuchElementException;

    boolean isPresent();

    boolean isEmpty();

    RegistryObject<T> or(RegistryObject<T> supplier);

    Identifier getId();

    T orElseGet(Supplier<T> supplier);

    <X extends Throwable> T orElseThrow(Supplier<X> supplier) throws X;

    Registry<T> getRegistry();
}

File: common/src/main/kotlin/org/sandboxpowered/silica/registry/RegistryObject.kt
package org.sandboxpowered.silica.registry

import org.sandboxpowered.silica.util.Identifier
import java.util.*
import java.util.function.Supplier

interface RegistryObject<T : RegistryEntry<T>> : Supplier<T> {
    fun asOptional(): Optional<T>

    @Throws(NoSuchElementException::class)
    override fun get(): T

    val isPresent: Boolean
    val isEmpty: Boolean
    fun or(supplier: RegistryObject<T>): RegistryObject<T>
    val id: Identifier
    fun orElseGet(supplier: Supplier<T>): T

    fun <X : Throwable> orElseThrow(supplier: Supplier<X>): T
    val registry: Registry<T>
}